From 0000000000000000000000000000000000000099 Mon Sep 17 00:00:00 2001
From: zhafknight <user-provided@local>
Date: Mon, 14 Sep 2026 11:44:00 +0800
Subject: [PATCH] frameworks/base: legacy low-RAM compatibility tweaks

Enable multi-window based on config even on low-RAM devices, extend the
SystemUI bind timeout to 30 seconds, and allow installation of apps that
target SDK levels below the platform minimum.

User-provided patch.
---
diff --git a/core/java/android/app/ActivityTaskManager.java b/core/java/android/app/ActivityTaskManager.java
index ad171a857959..24a12e725ab7 100644
--- a/core/java/android/app/ActivityTaskManager.java
+++ b/core/java/android/app/ActivityTaskManager.java
@@ -246,12 +246,8 @@ public class ActivityTaskManager {
     public static boolean supportsMultiWindow(Context context) {
         // On watches, multi-window is used to present essential system UI, and thus it must be
         // supported regardless of device memory characteristics.
-        boolean isWatch = context.getPackageManager().hasSystemFeature(
-                PackageManager.FEATURE_WATCH);
-        boolean forceOn = isWatch || ActivityManager.isMultiWindowForcedOn();
-        return (!ActivityManager.isLowRamDeviceStatic() || forceOn)
-                && Resources.getSystem().getBoolean(
-                com.android.internal.R.bool.config_supportsMultiWindow);
+    return Resources.getSystem().getBoolean(
+            com.android.internal.R.bool.config_supportsMultiWindow);
     }
 
     /**
diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index 143294982afa..b72fa76c8dcd 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -4866,8 +4866,13 @@ public class ActivityManagerService extends IActivityManager.Stub
 
             Message msg = mHandler.obtainMessage(BIND_APPLICATION_TIMEOUT_SOFT_MSG);
             msg.obj = app;
-            msg.arg1 = BIND_APPLICATION_TIMEOUT;
-            mHandler.sendMessageDelayed(msg, msg.arg1 /* BIND_APPLICATION_TIMEOUT */);
+            final int bindApplicationTimeout =
+                       "com.android.systemui".equals(app.processName)
+                               ? 30 * 1000 * Build.HW_TIMEOUT_MULTIPLIER
+                               : BIND_APPLICATION_TIMEOUT;
+
+            msg.arg1 = bindApplicationTimeout;
+            mHandler.sendMessageDelayed(msg, bindApplicationTimeout);
             mHandler.removeMessages(PROC_START_TIMEOUT_MSG, app);
 
             if (profilerInfo != null) {
diff --git a/services/core/java/com/android/server/pm/InstallPackageHelper.java b/services/core/java/com/android/server/pm/InstallPackageHelper.java
index 3d25f8f10175..5a121e42b3ca 100644
--- a/services/core/java/com/android/server/pm/InstallPackageHelper.java
+++ b/services/core/java/com/android/server/pm/InstallPackageHelper.java
@@ -24,7 +24,6 @@ import static android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DEFAULT;
 import static android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
 import static android.content.pm.PackageManager.INSTALL_FAILED_ALREADY_EXISTS;
 import static android.content.pm.PackageManager.INSTALL_FAILED_BAD_PERMISSION_GROUP;
-import static android.content.pm.PackageManager.INSTALL_FAILED_DEPRECATED_SDK_VERSION;
 import static android.content.pm.PackageManager.INSTALL_FAILED_DUPLICATE_PACKAGE;
 import static android.content.pm.PackageManager.INSTALL_FAILED_DUPLICATE_PERMISSION;
 import static android.content.pm.PackageManager.INSTALL_FAILED_DUPLICATE_PERMISSION_GROUP;
@@ -56,7 +55,6 @@ import static com.android.server.pm.PackageManagerService.DEBUG_PACKAGE_SCANNING
 import static com.android.server.pm.PackageManagerService.DEBUG_REMOVE;
 import static com.android.server.pm.PackageManagerService.DEBUG_UPGRADE;
 import static com.android.server.pm.PackageManagerService.DEBUG_VERIFY;
-import static com.android.server.pm.PackageManagerService.MIN_INSTALLABLE_TARGET_SDK;
 import static com.android.server.pm.PackageManagerService.PLATFORM_PACKAGE_NAME;
 import static com.android.server.pm.PackageManagerService.POST_INSTALL;
 import static com.android.server.pm.PackageManagerService.SCAN_AS_APEX;
@@ -1554,29 +1552,6 @@ final class InstallPackageHelper {
             Trace.traceEnd(TRACE_TAG_PACKAGE_MANAGER);
         }
 
-        // Block the install of apps using a lower target SDK version than required.
-        // This helps improve security and privacy as malware can target older SDK versions
-        // to avoid enforcement of new API behavior.
-        boolean bypassLowTargetSdkBlock =
-                ((installFlags & PackageManager.INSTALL_BYPASS_LOW_TARGET_SDK_BLOCK) != 0);
-
-        // Skip enforcement when the testOnly flag is set
-        if (!bypassLowTargetSdkBlock && parsedPackage.isTestOnly()) {
-            bypassLowTargetSdkBlock = true;
-        }
-
-        // Enforce the low target sdk install block except when
-        // the --bypass-low-target-sdk-block is set for the install
-        if (!bypassLowTargetSdkBlock
-                && parsedPackage.getTargetSdkVersion() < MIN_INSTALLABLE_TARGET_SDK) {
-            Slog.w(TAG, "App " + parsedPackage.getPackageName()
-                    + " targets deprecated sdk version");
-            throw new PrepareFailure(INSTALL_FAILED_DEPRECATED_SDK_VERSION,
-                    "App package must target at least SDK version "
-                            + MIN_INSTALLABLE_TARGET_SDK + ", but found "
-                            + parsedPackage.getTargetSdkVersion());
-        }
-
         // Instant apps have several additional install-time checks.
         if (instantApp) {
             if (parsedPackage.getTargetSdkVersion() < Build.VERSION_CODES.O) {

-- 
2.53.0
